### PR TITLE
[Parser] Lex floating point values

### PR DESCRIPTION
**Pull request by tlively**
Rather than trying to actually implement the parsing of float values, which
cannot be done naively due to precision concerns, just parse the float grammar
then postprocess the parsed text into a form we can pass to `strtod` to do the
actual parsing of the value.

Since the float grammar reuses `num` and `hexnum` from the integer grammar but
does not care about overflow, add a mode to `LexIntCtx`, `num`, and `hexnum` to
allow parsing overflowing numbers.

For NaNs, store the payload as a separate value rather than as part of the
parsed double. The payload will be injected into the NaN at a higher level of
the parser once we know whether we are parsing an f64 or an f32 and therefore
know what the allowable payload values are.

### CODE REVIEW

**Comment by tlively** [comment]
Current dependencies on/for this PR:
* main
  * **PR #4693** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/4693" target="_blank">Graphite</a>  👈
    * **PR #4694** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/4694" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/WebAssembly/binaryen/4693?utm_source=stack-comment).

File: src/wasm/wat-parser-internal.h
Comment (kripken) [inline src/wasm/wat-parser-internal.h:268]: The existing code uses `strtof` for an `f32`, but I'm not sure if that's necessary...

File: test/gtest/wat-parser.cpp
Comment (kripken) [inline test/gtest/wat-parser.cpp:834]: I'd say the spec tests may be good enough coverage for corner cases. But if you prefer to write out unit tests like this I'm not opposed.

File: src/wasm/wat-parser-internal.h
Comment (tlively) [reply src/wasm/wat-parser-internal.h:268]: There's no way to tell whether or not we are lexing an f32 or an f64 without higher level parser context, so we have to conservatively use double precision here. I don't think there are any problems that can arise from that.

File: test/gtest/wat-parser.cpp
Comment (tlively) [reply test/gtest/wat-parser.cpp:834]: We're a while off from being able to run spec tests with this new parser and the unit tests have been helpful for catching bugs.

File: src/wasm/wat-parser-internal.h
Comment (tlively) [reply src/wasm/wat-parser-internal.h:122]: I'll remove these and other unnecessary braces in a separate NFC PR.

Comment (kripken) [reply src/wasm/wat-parser-internal.h:122]: Sounds good.

File: test/gtest/wat-parser.cpp
Comment (kripken) [inline test/gtest/wat-parser.cpp:744]: Do we not need these NaN payloads for some spec tests?

Comment (kripken) [reply test/gtest/wat-parser.cpp:744]: I see, sounds good.

**Review by kripken — APPROVED** [review]

**Comment by tlively** [comment]
**Graphite Merge Job**

Current status: 🔄 Waiting for checks to pass

This pull request is currently waiting for CI and mergeability checks to pass before being merged.

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/WebAssembly/binaryen/4693?utm_source=stack-comment).


Job Reference: A9XO0urDPXHWLrhQjA6D

File: src/wasm/wat-parser-internal.h
Comment (aheejin) [inline src/wasm/wat-parser-internal.h:252]: Does this initialize base class members in order and then the child class members? Didn't know this was possible..

File: src/wasm/wat-parser-internal.h
Comment (aheejin) [inline src/wasm/wat-parser-internal.h:256]:
```suggestion
      return LexFloatResult{*basic, {}, -NAN};
```
Can `nanPayload` ever be non-null here, given that the `if` above does an early return?

File: src/wasm/wat-parser-internal.h
Comment (aheejin) [inline src/wasm/wat-parser-internal.h:822]: For non-NaN numbers, you store as a double conservatively and fix it later. Can't we do the same for NaNs?

File: test/gtest/wat-parser.cpp
Comment (aheejin) [inline test/gtest/wat-parser.cpp:740]: Do we not need payload for `nan`? The spec says `2^(signif(N)-1)`
```
'nan' => nan(2^(signif(N)-1))
```

File: test/gtest/wat-parser.cpp
Comment (aheejin) [inline test/gtest/wat-parser.cpp:825]: Aren't these `'nan:0x'n:hexnum` pattern?

File: test/gtest/wat-parser.cpp
Comment (aheejin) [inline test/gtest/wat-parser.cpp:713]: We don't recognize these as keywords? The same for `NaN`/`NAN`.

**Review by aheejin — COMMENTED** [review]
I guess most of the questions stem from my lack of knowledge about floating point representation, but anyway..

File: src/wasm/wat-parser-internal.h
Comment (tlively) [reply src/wasm/wat-parser-internal.h:252]: Yep, that's my understanding. I wouldn't have thought that this would work either, but apparently it does :shrug:

File: src/wasm/wat-parser-internal.h
Comment (tlively) [reply src/wasm/wat-parser-internal.h:256]: No, it will always be nullopt. I can make this change in a follow-up.

File: src/wasm/wat-parser-internal.h
Comment (tlively) [reply src/wasm/wat-parser-internal.h:822]: I thought about that, but the problem is that the default payload depends on the size of the float we are parsing. It's (2^52)-1 for f64 and (2^23)-1 for f32. If we stored the payload directly in the double to be fixed up later, then we wouldn't be able to tell the difference between a custom payload of (2^52)-1 that is out-of-bounds for an f32 and a default payload that was conservatively set to (2^53)-1 and can be fixed up. We could fix that in different ways, like having a bool `isDefaultPayload`, but this seems simpler overall.

File: test/gtest/wat-parser.cpp
Comment (tlively) [reply test/gtest/wat-parser.cpp:740]: Good observation. We will interpret a null payload to mean the default payload when `d` is NaN and inject that default payload in the same part of the parser (not implemented yet) where we will inject non-default payloads.

File: test/gtest/wat-parser.cpp
Comment (tlively) [reply test/gtest/wat-parser.cpp:825]: They would be, but `n` is outside the allowable bounds for f64 NaN payloads, so these fail to parse as floats. The lexer then tries to lex them as keywords, which succeeds. In the future once we have better error handling here, we should produce an error about the payload being out of bounds instead of falling back to parsing it as a keyword.

File: test/gtest/wat-parser.cpp
Comment (tlively) [reply test/gtest/wat-parser.cpp:713]: No, keywords can only start with lowercase letters :laughing:

File: test/gtest/wat-parser.cpp
Comment (aheejin) [reply test/gtest/wat-parser.cpp:740]: I see. Might be helpful for reading to add a comment after this block:
https://github.com/WebAssembly/binaryen/blob/590af8639ab4f6cfd82452b5ff2b431b470c64a5/src/wasm/wat-parser-internal.h#L602-L615

File: test/gtest/wat-parser.cpp
Comment (aheejin) [reply test/gtest/wat-parser.cpp:825]: We don't even allow `0x0`... 0 is also outside the allowable bounds?

Comment (tlively) [reply test/gtest/wat-parser.cpp:825]: That's right. Apparently if all the payload bits are zero, then the value is +-infinity rather than a NaN.

**Pull request merged**
